src/mpid/ch4/src/ch4_part.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+
+static MPIR_Request *part_req_create(void *buf, int partitions, MPI_Count count,
+                                     MPI_Datatype datatype, int rank, int tag,
+                                     MPIR_Comm * comm, MPIR_Request_kind_t kind)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *req = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_PART_REQ_CREATE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_PART_REQ_CREATE);
+
+    /* Set refcnt=1 for user-defined partitioned pattern; decrease at request_free. */
+    req = MPIR_Request_create_from_pool(kind, 0);
+    MPIR_ERR_CHKANDSTMT((req) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+
+    MPIR_Comm_add_ref(comm);
+    req->comm = comm;
+
+    MPIR_Datatype_add_ref_if_not_builtin(datatype);
+    MPIDI_PART_REQUEST(req, datatype) = datatype;
+
+    MPIDI_PART_REQUEST(req, rank) = rank;
+    MPIDI_PART_REQUEST(req, tag) = tag;
+    MPIDI_PART_REQUEST(req, context_id) = comm->context_id;
+    MPIDI_PART_REQUEST(req, buffer) = buf;
+    MPIDI_PART_REQUEST(req, count) = count;
+    MPIDI_PART_REQUEST(req, active_flag) = 0;
+    req->u.part.partitions = partitions;
+
+    /* Inactive partitioned comm request can be freed by request_free.
+     * Completion cntr increases when request becomes active at start. */
+    MPIR_cc_set(req->cc_ptr, 0);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPID_PART_REQ_CREATE);
+    return req;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPID_Psend_init(void *buf, int partitions, MPI_Count count,
+                    MPI_Datatype datatype, int dest, int tag,
+                    MPIR_Comm * comm, MPIR_Info * info, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_PSEND_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_PSEND_INIT);
+
+    MPIDI_av_entry_t *av = MPIDIU_comm_rank_to_av(comm, dest);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
+    MPIDI_workq_vci_progress_unsafe();
+
+    *request = part_req_create((void *) buf, partitions, count, datatype, dest, tag, comm,
+                               MPIR_REQUEST_KIND__PART_SEND);
+#ifdef MPIDI_CH4_DIRECT_NETMOD
+    mpi_errno = MPIDI_NM_mpi_psend_init(buf, partitions, count, datatype, dest, tag, comm,
```

Comment:
By the way, I think you should split this commit to separate ch4-part and mpidig part. The flow is `ch4 -> netmod -> mpidig`, so mpidig are not the same component level as the ch4 level here.

Reply:
The reason I put all ch4-layer code in a single commit is that, a part of the requests initialization is in `ch4_part.c`. Separating them will make the initialization logic incomplete. But as in the previous comment, we are discussing to move all initialization to `MPIDIG`. I may separate the ch4-part into a separate commit.